src/Arcus.Security.Core/Caching/CachedSecretProvider.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Arcus.Security.Core.Exceptions;
+using Arcus.Security.Core.Interfaces;
+using GuardNet;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace Arcus.Security.Core.Caching
+{
+    /// <inheritdoc cref="ISecretProvider"/>
+    /// <summary>
+    /// A Secret Provider that will cache secrets in memory, to improve performance
+    /// </summary>
+    public class CachedSecretProvider : ICachedSecretProvider
+    {
+        private readonly ISecretProvider _secretProvider;
+        private readonly IMemoryCache _memoryCache;
+        private readonly TimeSpan _cacheDuration;
+
+        /// <summary>
+        /// Creating a new CachedSecretProvider with all required information
+        /// </summary>
+        /// <param name="secretProvider">The internal <see cref="ISecretProvider"/> used to retrieve the actual Secret Value, when not cached</param>
+        /// <param name="cacheDuration">The <see cref="TimeSpan"/> for which a retrieved value should be cached</param>
+        /// <param name="memoryCache">A <see cref="IMemoryCache"/> implementation that can cache data in memory.</param>
+        /// <exception cref="ArgumentNullException">The secretProvider and memoryCache parameters must not be null</exception>
+        public CachedSecretProvider(ISecretProvider secretProvider, TimeSpan cacheDuration, IMemoryCache memoryCache)
+        {
+            Guard.NotNull(secretProvider, nameof(secretProvider));
+
+            _secretProvider = secretProvider;
+            _memoryCache = memoryCache ?? new MemoryCache(new MemoryCacheOptions());
```

Comment:
Maybe it's best to use overloads for this so that 1) our consumers know that it's possible to use it without their own cache 2) we can check for nulls if one is passed

Reply:
Done in new commit